File: PSO-粒子群算法(Particle Swarm Optimization)/MATLAB/PSO_3D.m
close all; clear; clc; 
%% Init Population
n = 3; N = 50;                  % 空间维数，种群个体个数
niter = 1000;                   % 迭代次数       
bnds = [0, 100;0, 100;0, 100];  % 取值范围  
vmin = -1; vmax = 1;            % 最小/大速度限
w = 0.8;                        % 惯性因子
c1 = 2; c2 = 2.1;               % 自我&群体学习因子
c = c1 + c2;
% K:收敛因子(保证收敛性, 通常令c1+c2=4.1, s.t. K=0.729)
K = 2 / (abs(2 - c - sqrt(c * c - 4 * c)));
% x: 个体初始取值
x = rand(N, n); y = rand(N, 1);
for i = 1:N
   for j = 1:n
      x(i, j) =  x(i, j) * (bnds(j, 2) - bnds(j, 1)) + bnds(j, 1);
   end
   y(i) = func(x(i,:));
end
v = rand(N, n);                 % 初始种群的速度
pbest_x = x;                    % 记录个体历史最优取值
pbest_y = y;                    % 记录个体历史最优值/适应度
gbest_y = min(y);               % 记录种群历史最优值/适应度
gbest_x = x(y == gbest_y, :);   % 记录种群历史最优取值
x0 = x; y0 = y;                 % 初始值备份
%% Main Loop
for l = 1:niter
   for i = 1:N
      for j = 1:n
          v(i, j) =  K * (v(i, j) + c1 * rand() * (pbest_x(i, j) - x(i, j)) + c2 * rand() * (gbest_x(j) - x(i, j)));
          v(i, j) = min(v(i, j), vmax); v(i, j) = max(v(i, j), vmin);
          x(i, j) = x(i, j) + v(i, j);       % 更新坐标
      end
      y(i) = func(x(i,:));           % 更新取值/适应度
      if y(i) < pbest_y(i)      % 更新个体最优记录
          pbest_x(i,:) = x(i,:);
          pbest_y(i) = y(i);
      end
      if y(i) < gbest_y         % 更新群体最优记录
          gbest_x = x(i,:);
          gbest_y = y(i);
      end
   end
end
fprintf('Global Minimum: xmin = (%f, %f, %f)\n', gbest_x);
fprintf('f(xmin) = %f\n', gbest_y);

File: PSO-粒子群算法(Particle Swarm Optimization)/PSO-3D.cpp
#include <iostream>
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <ctime>
#include <iomanip>
using namespace std;

/*
 * PSO: Particle Swarm Optimization Algorithm
 * 最好多运行几次取最优
 * Parameters:
 *  N: 种群个体数量(一般取20~50 or 100~200), niter: 迭代次数
 *  vmin, vmax: 最大最小速度--限定步长
 *  w: 惯性因子(>=0)
 *  c1, c2: 学习因子(通常c1=c2=2)
 *  K: 收敛因子(保证收敛性, 通常令c1+c2=4.1, s.t. K=0.729)
 *  bnds: 取值范围
 * 采用罚函数法将约束条件加入目标函数
 */


double getRand(){ return rand() / double(RAND_MAX);}


struct Points{
    double x[3] = {getRand(), getRand(), getRand()};
    int size = 3;
};

void print(Points x){
    for (int i = 0; i < x.size; ++i)
        cout << x.x[i] << (i == x.size - 1?"\n":" ");
    return;
}

double target_func(Points p){return p.x[0] * p.x[0] + p.x[1] * p.x[1] + p.x[2] * p.x[2] + 8;}

// 若无约束则可省略惩罚步骤
double f(Points p){ // 违背约束条件则惩罚
    double y = target_func(p), penalty = 0x3f3f3f3f, bnds;
    if (-p.x[0] > 0){y += (penalty * (-p.x[0]));} // -x1 <= 0
    if (-p.x[1] > 0){y += (penalty * (-p.x[1]));} // -x2 <= 0
    if (-p.x[2] > 0){y += (penalty * (-p.x[2]));} // -x3 <= 0
    bnds = -p.x[0] * p.x[0] + p.x[1] - p.x[2] * p.x[2]; // -x1^2 + x2 - x3^2 <= 0
    if (bnds > 0){y += (penalty * bnds);}
    bnds = p.x[0] + p.x[1] * p.x[1] + p.x[2] * p.x[2] - 20; // x1 + x2^2 + x3^2 - 20 <= 0
    if (bnds > 0){y += (penalty * bnds);}
    bnds = -p.x[0] - p.x[1] * p.x[1] + 2; // -x1 - x2^2 + 2 <= 0
    if (bnds > 0){y += (penalty * bnds);}
    bnds = -p.x[1] - 2 * p.x[2] * p.x[2] + 3; // -x2 - 2 * x3^2 + 3 <= 0
    if (bnds > 0){y += (penalty * bnds);}
    return y;
}


const int N = 50, niter = 500;
double vmin = -1, vmax = 1;
double w = 1, c1 = 2, c2 = 2.1;
double c = c1 + c2;
double K = 2 / (abs(2 - c - sqrt(c * c - 4 * c)));
Points x[N], v[N], pbest_x[N], gbest_x, vi;
double y[N], pbest_y[N], gbest_y;
void PSO(){
    // init
    for (int i = 0; i < N; ++i) {y[i] = pbest_y[i] = f(x[i]);}
    gbest_x = x[0]; gbest_y = y[0];
    for (int i = 0; i < N; ++i) {
        if (y[i] < gbest_y){
            gbest_y = y[i];
            gbest_x = x[i];
        }
    }
    // main loop
    for (int l = 0; l < niter; ++l) {
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < vi.size; ++j) {
                vi.x[j] = K * (v[i].x[j] + c1 * getRand()
                        * (pbest_x[i].x[j] - x[i].x[j]) + c2 * getRand() * (gbest_x.x[j] - x[i].x[j]));
                vi.x[j] = min(vi.x[j], vmax);
                vi.x[j] = max(vi.x[j], vmin);
            }
            for (int j = 0; j < x[i].size; ++j) {
                x[i].x[j] += vi.x[j];
            }
            y[i] = f(x[i]);
            if (y[i] < pbest_y[i]){
                pbest_x[i] = x[i];
                pbest_y[i] = y[i];
            }
            if (y[i] < gbest_y){
                gbest_x = x[i];
                gbest_y = y[i];
            }
        }
    }
    cout << "Global Minimum: xmin = ";
    print(gbest_x);
    cout << "f(xmin) = " << gbest_y << endl;
    return;
}



int main()
{
    ios_base::sync_with_stdio(false);
    time_t s, e;
    srand((unsigned int)time(0));
    time(&s);
    PSO();
    time(&e);
    cout << "Running Time: " << e - s << "s" << endl;
    return 0;
}

/*
 * Global Minimum: xmin = 0.615724 1.1783 0.964955
 * f(xmin) = 10.6986
 * Running Time: 0s
 */
